tools/appveyor.psm1

```diff
@@ -0,0 +1,201 @@
+$ErrorActionPreference = 'Stop'
+$repoRoot = Join-Path $PSScriptRoot '..'
+
+Import-Module (Join-Path $repoRoot 'build.psm1')
+
+# Emulates running all of AppVeyor but locally
+# should not be used on AppVeyor
+function Invoke-AppVeyorFull
+{
+    Invoke-AppVeyorInstall
+    Invoke-AppVeyorBuild
+    Invoke-AppVeyorTest
+    Invoke-AppveyorFinish
+}
+
+# Implements the AppVeyor 'build_script' step
+function Invoke-AppVeyorBuild
+{
+      # check to be sure our test tags are correct
+      $result = Get-PesterTag
+      if ( $result.Result -ne "Pass" ) {
+        $result.Warnings
+        throw "Tags must be CI, Feature, Scenario, or Slow"
+      }
+      Start-PSBuild -CrossGen -Configuration $buildConfiguration
+      Start-PSBuild -FullCLR
+}
+
+# Implements the AppVeyor 'install' step
+function Invoke-AppVeyorInstall
+{
+    Start-PSBootstrap -Force
+}
+
+# A wrapper to ensure that we upload test results
+# and that if we are not able to that it does not fail
+# the CI build
+function Update-AppVeyorTestResults
+{
+    param(
+        [string] $resultsFile
+    )
+
+    if($env:Appveyor)
+    {
+        $retryCount = 0 
+        $pushedResults = $false
+        $pushedArtifacts = $false 
+        while( (!$pushedResults -or !$pushedResults) -and $retryCount -lt 3)
+        {
+            if($retryCount -gt 0)
+            {
+                Write-Verbose "Retrying updating test artifacts..."
+            }
+
+            $retryCount++
+            $resolvedResultsPath = (Resolve-Path $resultsFile)
+            try {
+                (New-Object 'System.Net.WebClient').UploadFile("https://ci.appveyor.com/api/testresults/nunit/$($env:APPVEYOR_JOB_ID)", $resolvedResultsPath)
+                $pushedResults = $true
+            }
+            catch {
+                Write-Warning "Pushing test result failed..."
+            }
+
+            try {
+                Push-AppveyorArtifact $resolvedResultsPath
+                $pushedArtifacts = $true
+            }
+            catch {
+                Write-Warning "Pushing test Artifact failed..."
+            }
+        }
+
+        if(!$pushedResults -or !$pushedResults)
+        {
+            Write-Warning "Failed to push all artifacts for $resultsFile"
+        }
+    }
+    else 
+    {
+        Write-Warning "Not running in appveyor, skipping upload of test results: $resultsFile"
+    }
+}
+
+# Implement AppVeyor 'Test_script'
+function Invoke-AppVeyorTest 
+{
+    #
+    # CoreCLR
+    $env:CoreOutput = Split-Path -Parent (Get-PSOutput -Options (New-PSOptions -Publish -Configuration $buildConfiguration))
+    Write-Host -Foreground Green 'Run CoreCLR tests'
+    $testResultsFile = "$pwd\TestsResults.xml"
+    if(!(Test-Path "$env:CoreOutput\powershell.exe"))
+    {
+        throw "CoreCL PowerShell.exe was not built"
```

Comment:
Resolved

---
In reply to: [76693422](https://github.com/PowerShell/PowerShell/pull/2121#discussion_r76693422) [](ancestors = 76693422)